include/flatbuffers/flatbuffers.h

```diff
@@ -542,57 +563,228 @@ struct String : public Vector<char> {
+ public:
+  virtual ~Allocator() {}
+
+  // Allocate `size` bytes of memory.
+  virtual uint8_t *allocate(size_t size) = 0;
+
+  // Deallocate `size` bytes of memory at `p` allocated by this allocator.
+  virtual void deallocate(uint8_t *p, size_t size) = 0;
+
+  // Reallocate `new_size` bytes of memory, replacing the old region of size
+  // `old_size` at `p`. In contrast to a normal realloc, this grows downwards,
+  // and is intended specifcally for `vector_downward` use.
+  virtual uint8_t *reallocate_downward(uint8_t *old_p,
+                                       size_t old_size,
+                                       size_t new_size) {
+    assert(new_size > old_size);  // vector_downward only grows
+    uint8_t *new_p = allocate(new_size);
+    memcpy(new_p + (new_size - old_size), old_p, old_size);
+    deallocate(old_p, old_size);
+    return new_p;
+  }
+};
+
+// DefaultAllocator uses new/delete to allocate memory regions
+class DefaultAllocator : public Allocator {
+ public:
+  virtual uint8_t *allocate(size_t size) FLATBUFFERS_OVERRIDE {
+    return new uint8_t[size];
+  }
+
+  virtual void deallocate(uint8_t *p, size_t) FLATBUFFERS_OVERRIDE {
+    delete[] p;
+  }
+};
+
+// simple_allocator is retained for backwards compatibility, but we should
+// eventually deprecate and remove it.
+class simple_allocator : public DefaultAllocator {
  public:
   virtual ~simple_allocator() {}
-  virtual uint8_t *allocate(size_t size) const { return new uint8_t[size]; }
-  virtual void deallocate(uint8_t *p) const { delete[] p; }
+  virtual uint8_t *allocate(size_t size) {
+    return new uint8_t[size];
+  }
+  virtual void deallocate(uint8_t *p) {
+    delete[] p;
+  }
+  virtual void deallocate(uint8_t *p, size_t) FLATBUFFERS_OVERRIDE {
+    deallocate(p);
+  }
+};
+
+// AllocatorProxy references an externally-managed allocator
+class AllocatorProxy : public Allocator {
+ public:
+  inline AllocatorProxy(Allocator *allocator) : allocator_(allocator) {}
+
+  inline AllocatorProxy(const AllocatorProxy &other)
+    : allocator_(other.allocator_) {}
+
+  inline AllocatorProxy &operator=(const AllocatorProxy &other) {
+    allocator_ = other.allocator_;
+    return *this;
+  }
+
+  virtual inline uint8_t *allocate(size_t size) FLATBUFFERS_OVERRIDE {
+    return allocator_->allocate(size);
+  }
+
+  virtual inline void deallocate(uint8_t *p, size_t size) FLATBUFFERS_OVERRIDE {
+    allocator_->deallocate(p, size);
+  }
+
+  virtual uint8_t *reallocate_downward(
+      uint8_t *old_p, size_t old_size, size_t new_size) FLATBUFFERS_OVERRIDE {
+    return allocator_->reallocate_downward(old_p, old_size, new_size);
+  }
+
+ protected:
+  Allocator *allocator_;
+};
+
+// DetachedBuffer is specific to the `release` functionality, and allows the
+// user to safely transfer ownership of a buffer. This resembles a unique_ptr +
+// custom deleter in many ways, but we do it here explicitly for wider compiler
+// support.
+class DetachedBuffer {
```

Comment:
We haven't had problems with `unique_ptr` on VS2010 (==1600) before (including the code you're replacing), so not sure what the problem is here?
We do run the full set of tests on VS2010, which needs this functionality, so we cannot define for CPP98_STL for VS2010.